### PR TITLE
Add HPA In-Context

### PR DESCRIPTION
**Pull request by andrewballantyne**
**Fixes**: 
https://issues.redhat.com/browse/ODC-4269

**Analysis / Root cause**: 
As a user, I'd like to be able to add an HPA to my D or DC node from Topology.

**Solution Description**: 
Add a kebab option to the D/DC nodes that redirects to an 'in-context' add HPA page.

**Screen shots / Gifs for design review**: 

@openshift/team-devconsole-ux 
@lwrigh 

Deployment with no limits (On Form)
Screen Shot 2020-07-17 at 1 56 13 PM

Deployment with no limits (On YAML)
Screen Shot 2020-07-17 at 1 56 36 PM

DeploymentConfig with CPU Limits:
Screen Shot 2020-07-17 at 1 58 27 PM

Can't have less than 1 Min Pod / Can't have more Max than Min Pods:
Screen Shot 2020-07-24 at 10 55 23 AM

Modifying the Utilization to be a non-number on YAML and trying to submit on the Form:
Screen Shot 2020-07-17 at 1 59 00 PM

Modifying a metric for memory (on YAML) without Memory Limits and trying to submit on Form:
Screen Shot 2020-07-17 at 1 59 52 PM

**Unit test coverage report**: 

```
 PASS  packages/dev-console/src/components/hpa/__tests__/hpa-utils.spec.ts
  isCpuUtilizationPossible provides accurate checks
    ✓ expect an invalid resource to return no (2ms)
    ✓ expect a resource with no limits to be not possible
    ✓ expect a resource with other limits but not CPU limits to be not possible
    ✓ expect a resource with CPU limits to be possible
    ✓ expect a resource with both CPU And Memory limits to be possible
  isMemoryUtilizationPossible provides accurate checks
    ✓ expect an invalid resource to return no (1ms)
    ✓ expect a resource with no limits to be not possible
    ✓ expect a resource with other limits but not CPU limits to be not possible
    ✓ expect a resource with CPU limits to be possible
    ✓ expect a resource with both CPU And Memory limits to be possible
  getLimitWarning provides a string when limits are lacking
    ✓ expect invalid resource to return a value
    ✓ expect no limit resources to return a value (1ms)
    ✓ expect partial limit resources to return a value
    ✓ expect full limits to not return a value
  getFormData gets back an hpa structured form object
    ✓ expect to be scaled against the resource provided (4ms)
    ✓ expect to be able to build off an existing HPA (1ms)
  getYAMLData gets back an hpa structured editor string
    ✓ expect to be scaled against the resource provided (2ms)
    ✓ expect to be able to build off an existing HPA
  getMetricByType returns an appropriate metric and the index it is at
    ✓ expect no metrics to return a default metric as a new metric on the end (1ms)
    ✓ expect to get back a default memory metric when only cpu metric is available
    ✓ expect to get back the cpu metric when it is available
  sanitizeHPAToForm always returns valid HPA
    ✓ expect an empty form to return a default HPA (2ms)
    ✓ expect it not to remove existing metrics (1ms)
  sanityForSubmit covers some basic field locking and trimming
    ✓ expect to always return in the deployment namespace
    ✓ expect to always scale to the same resource despite hpa settings (1ms)
    ✓ expect not to have empty cpu or memory metrics
    ✓ expect not to trim custom resource metrics
  getInvalidUsageError returns an error string when limits are not set
    ✓ expect cpu metric not to be allowed while disabled
    ✓ expect memory metric to not be allowed while disabled

Test Suites: 1 passed, 1 total
Tests:       29 passed, 29 total

------------------------------------------------------------|----------|----------|----------|----------|-------------------|
File                                                        |  % Stmts | % Branch |  % Funcs |  % Lines | Uncovered Line #s |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
  src/components/hpa/hpa-utils.ts                           |    96.97 |    90.32 |      100 |    96.49 |            92,100 |
------------------------------------------------------------|----------|----------|----------|----------|-------------------|
```

**Test setup:**

- Have a D or DC
- Add via the Topology Sidebar
- You can set limits on your D/DC by using the Advanced Option "Resource Limits" on the Add/Edit Add forms

**Browser conformance**: 
- [x] Chrome
- [ ] Firefox
- [ ] Safari
- [ ] Edge

### CODE REVIEW

File: frontend/packages/console-shared/src/components/formik-fields/SyncedEditorField.tsx [hunk 1]
```diff
@@ -12,6 +12,7 @@ type EditorContext = {
   name: string;
   editor: React.ReactNode;
   isDisabled?: boolean;
+  sanitizeTo?: (preFormData: any) => any;
```
Comment (andrewballantyne) [inline frontend/packages/console-shared/src/components/formik-fields/SyncedEditorField.tsx:15]: @rohitkrai03 I added this - let me know if you have any concerns about this functionality. I needed a way to get my root data back to a stable state when going from YAML => Form.

**Comment by andrewballantyne** [comment]
Need to still add tests, but this can be reviewed.

**Comment by lwrigh** [comment]
All of the behaviors and screenshots are looking great! My only comments are that I would fully spell out minimum or maximum pods in the form validation error messages. In the yellow alert message, does the message wrap past 3 lines if the alert box is the same width as the other fields? I was hoping that we could keep everything close to the same width. @andrewballantyne

**Comment by andrewballantyne** [comment]
> All of the behaviors and screenshots are looking great! My only comments are that I would fully spell out minimum or maximum pods in the form validation error messages. In the yellow alert message, does the message wrap past 3 lines if the alert box is the same width as the other fields? I was hoping that we could keep everything close to the same width. @andrewballantyne

@lwrigh We could definitely try to keep the content to the same width... but the YAML takes up the whole screen unless the sidebar is open; of which I just noticed we do not have haha... not sure why that's not there. I'll have to investigate that. I kept the content of the form smaller as the input fields looked kinda goofy spreading out when the pod spinners did not.

The help text changes are a good catch, guess I was hacking that out and didn't really think too hard on it.

**Comment by debsmita1** [comment]
Tried this out locally, `maxReplicas` is a required field & it can be greater or equal to `minReplicas`

File: frontend/packages/dev-console/src/plugin.tsx [hunk 2]
```diff
@@ -804,6 +804,16 @@ const plugin: Plugin<ConsumedExtensions> = [
         ).default,
     },
   },
+  {
+    type: 'Page/Route',
+    properties: {
+      exact: true,
+      path: ['/workload-hpa/:ns/:kind/:name'],
```
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/plugin.tsx:811]: I opted out of this to align with our custom pages like Dev Monitoring and Helm Releases... It's a custom in-context page for dev console, so I chose to go with not a k8s routed page.

**Comment by andrewballantyne** [comment]
> Tried this out locally, `maxReplicas` is a required field & it can be greater or equal to `minReplicas`

I'll update the error text... as that is indeed true. One can set min/max the same - although I'm not sure you'd want to use an HPA if you didn't allow it to scale.

File: frontend/packages/console-shared/src/components/formik-fields/SyncedEditorField.tsx (hunk 1 above)
Comment (rohitkrai03) [reply frontend/packages/console-shared/src/components/formik-fields/SyncedEditorField.tsx:15]: Yeah, no problem. This looks good to me.

**Comment by andrewballantyne** [comment]
Updated wording to spell out max and min verbiage and to include the fact that they can be equal to each other. The sidebar will have to come in another PR... it's not supported by the YAML editor that is used by the form/yaml switcher.

**Comment by debsmita1** [comment]
We would need to add proper validation for the name field. We can re-use what we have for the name field in the git import form
namevalidation

Also,
hpa

File: frontend/packages/dev-console/src/plugin.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (debsmita1) [reply frontend/packages/dev-console/src/plugin.tsx:811]: Then the route should be
```suggestion
      path: ['/workload-hpa/ns/:ns/:kind/:name'],
```

**Comment by andrewballantyne** [comment]
> We would need to add proper validation for the name field. We can re-use what we have for the name field in the git import form
> namevalidation

Ah good catch/call... I can add that `yup` validation!

> Also,
> hpa

Ah crud, forgot I put that fix in the Edit PR and not this one. It's fixed in #6039

File: frontend/packages/dev-console/src/plugin.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/plugin.tsx:811]: Not sure I can really argue with that one... I think it doesn't really conform well to the way a lot of our pages work, but that's a simple enough change for me to do.

**Comment by serenamarie125** [comment]
@andrewballantyne I find it a bit odd that our inline notifications aren't all the same width. Thoughts @lwrigh @beaumorley ?

**Comment by andrewballantyne** [comment]
> @andrewballantyne I find it a bit odd that our inline notifications aren't all the same width. Thoughts @lwrigh @beaumorley ?

Yeah... the footer takes up the space it takes. The form is a fixed width for the fields... and the top message is unbounded. I suppose I could make the inline info in the form take up full width... but I don't think we should touch the footer message. And I'm unsure if we need to really constrain the top-level messages.

File: frontend/packages/dev-console/src/plugin.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/plugin.tsx]: Updated

**Comment by andrewballantyne** [comment]
@beaumorley @lwrigh @serenamarie125 I made all the infos the same length, let me know if this is okay.

image

**Comment by andrewballantyne** [comment]
@debsmita1 please do another review.

File: frontend/packages/dev-console/src/components/hpa/HPADetailsForm.tsx [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+import { InputField, NumberSpinnerField } from '@console/shared';
+import { HorizontalPodAutoscalerKind } from '@console/internal/module/k8s';
+import HPAUtilizationField from './HPAUtilizationField';
+import { getMetricByType } from './hpa-utils';
+import { HPAFormValues, SupportedMetricTypes } from './types';
+
+const HPADetailsForm: React.FC = () => {
+  const name = 'formData';
+  const [field] = useField<HorizontalPodAutoscalerKind>(name);
+  const { setFieldValue, values } = useFormikContext<HPAFormValues>();
```
Comment (debsmita1) [inline frontend/packages/dev-console/src/components/hpa/HPADetailsForm.tsx:13]:
```suggestion
  const {
    setFieldValue,
    values: {
      showCanUseYAMLMessage,
      disabledFields: { cpuUtilization, memoryUtilization },
    },
  } = useFormikContext<HPAFormValues>();
```

**Comment by debsmita1** [comment]
If the form has some errors and I switch to YAML editor and fix it. on switching back to form view I still see those field errors. Have a look at the below gif:
hpa

**Comment by andrewballantyne** [comment]
> Should we still show the helptext `CPU request and limit must be set before CPU utilization can be set` when the CPU utilization field is editable? same for memory utilization field

Hmmm, that shouldn't be possible. I'll have to look into that... pretty sure they have the same check so it shouldn't be possible to have the warning & not be disabled.

File: frontend/packages/dev-console/src/components/hpa/HPADetailsForm.tsx (hunk 3 above)
Comment (andrewballantyne) [reply frontend/packages/dev-console/src/components/hpa/HPADetailsForm.tsx:13]: Yeah, that's a good idea 👍

**Comment by andrewballantyne** [comment]
@debsmita1 I can't reproduce the warning being there and the field being enabled. Please let me know if you can reproduce it. They used the same logic, so I am unsure how it could happen 🤔 

I've also fixed the two other things you mentioned, PTAL.

**Comment by debsmita1** [comment]
pulled in your latest changes and it works fine. Thank you!
/lgtm

**Pull request merged**